crates/chain-state/src/in_memory.rs

```diff
@@ -417,6 +417,22 @@ impl CanonicalInMemoryState {
 
         MemoryOverlayStateProvider::new(in_memory, historical)
     }
+
+    /// Returns an iterator over all canonical blocks in the in-memory state, from newest to oldest.
+    pub fn canonical_chain(&self) -> impl Iterator<Item = Arc<BlockState>> {
+        let pending = self.inner.in_memory_state.pending.read().clone();
+        let head = self.inner.in_memory_state.head_state();
+
+        let blocks = self.inner.in_memory_state.blocks.read().clone();
```

Comment:
this should be okay because we only expect a few blocks and all of them are just arcs